pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -98,32 +109,48 @@ def objective(self, budgets: list[float]) -> float:
             The negative total response value.
 
         """
-        total_response = 0
         first_transform, second_transform = (
             (self.adstock, self.saturation)
             if self.adstock_first
             else (self.saturation, self.adstock)
         )
-        for idx, (_channel, params) in enumerate(self.parameters.items()):
-            budget = budgets[idx] / self.scales[idx]
-            first_params = (
-                params["adstock_params"]
-                if self.adstock_first
-                else params["saturation_params"]
-            )
-            second_params = (
-                params["saturation_params"]
-                if self.adstock_first
-                else params["adstock_params"]
-            )
-            spend = np.full(self.num_periods, budget)
-            spend_extended = np.concatenate([spend, np.zeros(self.adstock.l_max)])
-            transformed_spend = second_transform.function(
-                x=first_transform.function(x=spend_extended, **first_params),
-                **second_params,
-            ).eval()
-            total_response += np.sum(transformed_spend)
-        return -total_response
+
+        budget = budgets / self.scales  # dim: (channels)
+
+        first_params = (
+            self.parameters["adstock_params"]
+            if self.adstock_first
+            else self.parameters["saturation_params"]
+        )
+        second_params = (
+            self.parameters["saturation_params"]
+            if self.adstock_first
+            else self.parameters["adstock_params"]
+        )
+
+        spend = np.tile(budget, (self.num_periods, 1))  # dim: (periods, channels)
+        spend_extended = np.vstack(
+            [spend, np.zeros((self.adstock.l_max, spend.shape[1]))]
+        )  # dim: (periods + l_max, channels)
+
+        _response = first_transform.function(x=spend_extended, **first_params)
+
+        # Dynamically reshape parameters in `second_params` to match `_response`
+        for param_name, param_value in second_params.items():
+            if param_value.ndim == 3:  # Check if it lacks the `time` dimension
+                # Reshape by adding a singleton dimension for broadcasting over time
+                second_params[param_name] = param_value.reshape(
+                    param_value.shape[0], param_value.shape[1], 1, param_value.shape[2]
+                )
```

Comment:
Can we make an issue for this if our plan is to extend.

Reply:
Sure, let me open an issue!